File: 200. Number of Islands.swift
class Solution {
    var islandCount = 0
    
    var tempGrid = [[Character]]()
    
    func numIslands(_ grid: [[Character]]) -> Int {
        
        if grid.count == 0 || grid[0].count == 0 {
            return 0
        }
        
        tempGrid = grid
        
        // loop all node
        for x in 0 ..< tempGrid.count {
            for y in 0 ..< tempGrid[0].count {
                if tempGrid[x][y] == "1" {
                    markByBfs(&tempGrid, x, y)
                    islandCount += 1
                }
            }
        }
        
        return islandCount
    }
    
    struct Coordinate {
        var x: Int
        var y: Int
    }
    
    // bfs to check islands
    func markByBfs(_ grid: inout [[Character]], _ x: Int, _ y: Int) {
        // magic numbers!
        let directionX = [0, 1, -1, 0]
        let directionY = [1, 0, 0, -1]
        
        var queue = Queue<Coordinate>()
        queue.enqueue(Coordinate(x: x, y: y))
        
        while !queue.isEmpty {
            if let pop = queue.dequeue() {
                for i in 0 ..< directionX.count {
                    let adjPointX = pop.x + directionX[i]
                    let adjPointY = pop.y + directionY[i]
                    if isOutBound(adjPointX, adjPointY) {
                        continue
                    }
                    
                    if grid[adjPointX][adjPointY] == "1" {
                        queue.enqueue(Coordinate(x: adjPointX, y: adjPointY))
                        grid[adjPointX][adjPointY] = "0"
                    }
                }
            }
        }
    }
    
    func isOutBound(_ x: Int, _ y: Int) -> Bool {
        if x < 0 || x >= tempGrid.count || y < 0 || y >= tempGrid[0].count {
            return true
        }
        
        return false
    }
}

public struct Queue<T> {
    fileprivate var array = [T]()
    
    public var isEmpty: Bool {
        return array.isEmpty
    }
    
    public var count: Int {
        return array.count
    }
    
    public mutating func enqueue(_ element: T) {
        array.append(element)
    }
    
    public mutating func dequeue() -> T? {
        if isEmpty {
            return nil
        } else {
            return array.removeFirst()
        }
    }
    
    public var front: T? {
        return array.first
    }
}

File: LeetCode/200.Number of Islands.py
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        if len(grid) == 0 or len(grid[0]) == 0:
            return 0
        m, n = len(grid), len(grid[0])
        mark = map(lambda x: map(lambda y: True if y == '1' else False, x) , grid)
        count = 0
        for i in range(m):
            for j in range(n):
                if mark[i][j]:
                    count += 1
                    mark[i][j] = False
                    record = [(i, j)]
                    while len(record) > 0:
                        x, y = record.pop()
                        if x + 1 < m and mark[x + 1][y]:
                            mark[x + 1][y] = False
                            record.append((x + 1, y))
                        if x - 1 >= 0 and mark[x - 1][y]:
                            mark[x - 1][y] = False
                            record.append((x - 1, y))
                        if y + 1 < n and mark[x][y + 1]:
                            mark[x][y + 1] = False
                            record.append((x, y + 1))
                        if y - 1 >= 0 and mark[x][y - 1]:
                            mark[x][y - 1] = False
                            record.append((x, y - 1))
        return count
